### PR TITLE
Code cleanup + Bugfix

### PR DESCRIPTION
**Pull request by JonhasSC**
This PR is focused on cleaning up some of the code, specifically, the interfaces for generating resource estimates. Additionally, this PR is focused on fixing up some bugs w/ respect to GSEE that was uncovered after PR https://github.com/lanl-ansi/qc-applications/pull/61 was merged in. The bugs just broke the handshake between circuit generation + resource estimation so that was handled as well.

### CODE REVIEW

**Comment by gsgrattan** [comment]
WIP Review

**Review by gsgrattan — APPROVED** [review]
A few minor questions but otherwise looks good!

File: src/qca/utils/utils.py [hunk 1]
```diff
@@ -337,3 +350,35 @@ def re_as_json(main_estimate:dict, outdir:str) -> None:
         json.dump(main_estimate, f,
             indent=4,
             separators=(',', ': '))
+
+def gen_json(main_estimate: dict, outfile:str, metadatata: EstimateMetaData|None=None):
+    if metadatata:
+        re_metadata = asdict(metadatata)
```
[the remark this answers is not in the document]
Comment (JonhasSC) [reply src/qca/utils/utils.py:356]: asdict transforms a dataclass into a dictionary. If it exists, the dictionary that gets output from here will always exist

File: src/qca/utils/algo_utils.py [hunk 2]
```diff
@@ -219,23 +220,23 @@ def gsee_resource_estimation(
...
+        is_extrapolated = True
+
+    grab_circuit_resources(
         circuit=pe_circuit,
         outdir=outdir,
-        numsteps=nsteps,
         algo_name='GSEE',
+        fname=circuit_name,
+        is_extrapolated=is_extrapolated,
+        numsteps=nsteps,
```
[the remark this answers is not in the document]
Comment (JonhasSC) [reply src/qca/utils/algo_utils.py:234]: total_steps probably best

File: src/qca/utils/utils.py [hunk 3]
```diff
@@ -153,48 +157,51 @@ def gen_resource_estimate(
...
+    scaling_factor = 0
+    if resource != 0:
+        scaled_steps = 0
+        scaled_bits = 0
+        if total_steps and total_steps > 0:
+            scaled_steps = total_steps
+        if bits_precision and bits_precision > 0:
+            scaled_bits = pow(2, bits_precision - 1)
+        
+        if scaled_steps and scaled_bits:
```
[the remark this answers is not in the document]
Comment (JonhasSC) [reply src/qca/utils/utils.py:175]: In python, you are allowed to express a boolean expression through the numerical value of the variable. Thus, if scaled_steps or scaled_bits never get reassigned, they will never evaluate these branches. For example, in your terminal, if you do the following:
```python
x =0
if x:
    print('reached this branch')
```
you will see that this branch will never evaluate to true and thus, never be taken. In this example, if x is any other number besides 0, negative or positive, the branch will ALWAYS be evaluated.

File: src/qca/utils/utils.py [hunk 4]
```diff
@@ -153,48 +157,51 @@ def gen_resource_estimate(
...
-        resource_estimate['circuit_depth'] = resource_estimate['circuit_depth'] * scaling_factor
-        resource_estimate['clifford_count'] = resource_estimate['clifford_count'] * scaling_factor
-
-    if circ_occurences > 0:
+    if circ_occurences:
         resource_estimate['subcircuit_occurences'] = circ_occurences
 
     return resource_estimate
 
+def scale_resource(resource: int, total_steps: int | None=None, bits_precision: float | None = None) -> int:
```
[the remark this answers is not in the document]
Comment (JonhasSC) [reply src/qca/utils/utils.py:165]: The idea behind this function is that if nsteps or bits_precisions are both evaluated to None, then this is inherently a NOP operation. Im not sure what you mean about invalid or missing parameters tho

**Review by zain2864 — APPROVED** [review]
Test cases for scale_resource looks good to me, thanks!

**Pull request merged**